tools/clang/lib/SPIRV/RawBufferMethods.cpp

```diff
@@ -454,24 +463,25 @@ SpirvInstruction *RawBufferHandler::processTemplatedLoadFromBuffer(
   if (const auto *structType = targetType->getAs<RecordType>()) {
     const auto *decl = structType->getDecl();
     assert(bitOffset == 0);
-    const auto &layout = astContext.getASTRecordLayout(decl);
     SpirvInstruction *originalIndex = index;
     uint32_t originalBitOffset = bitOffset;
     llvm::SmallVector<SpirvInstruction *, 4> loadedElems;
-    uint32_t structAlignment = 0;
-    uint32_t structSize = static_cast<uint32_t>(layout.getSize().getQuantity());
-    uint32_t fieldIndex = 0;
+    uint32_t fieldOffsetInBytes = 0;
+    uint32_t structAlignment = 0, structSize = 0, stride = 0;
+    std::tie(structAlignment, structSize) =
+        AlignmentSizeCalculator(astContext, theEmitter.getSpirvOptions())
+            .getAlignmentAndSize(targetType,
+                                 theEmitter.getSpirvOptions().sBufferLayoutRule,
+                                 llvm::None, &stride);
     for (const auto *field : decl->fields()) {
-      CharUnits alignment;
-      std::tie(std::ignore, alignment) = astContext.getTypeInfoInChars(
-          field->getType()->getUnqualifiedDesugaredType());
-      structAlignment = std::max(
-          structAlignment, static_cast<uint32_t>(alignment.getQuantity()));
-
-      assert(fieldIndex < layout.getFieldCount());
-      uint32_t fieldOffsetInBytes = static_cast<uint32_t>(
-          astContext.toCharUnitsFromBits(layout.getFieldOffset(fieldIndex++))
-              .getQuantity());
+      AlignmentSizeCalculator alignmentCalc(astContext,
+                                            theEmitter.getSpirvOptions());
+      uint32_t fieldSize = 0, fieldAlignment = 0;
+      std::tie(fieldAlignment, fieldSize) = alignmentCalc.getAlignmentAndSize(
```

Comment:
Based on my understanding, `getAlignmentAndSize()` looks fine but I am worried it is a little bit complicated. I am curious we have unittests to test only this small function. Actually, we have several complicated parts of code and if we tested them using unittests, it would be nice. Let's discuss this issue in offline because this is probably not directly related to this PR.

Reply:
Sure, sounds good to me.